File: LeetCodeProblemsGo/LeetCode1399CountLargestGroup/LeetCode1399CountLargestGroup.go
package LeetCode1399CountLargestGroup

func countLargestGroup(n int) int {
	digitMap := make(map[int]int)
	digitSum := 0
	maxGroupSize := 0
	for i := 1; i <= n; i++ {
		digitSum = getCount(i)
		if _, exists := digitMap[digitSum]; !exists {
			digitMap[digitSum] = 0
		}
		digitMap[digitSum]++
		if digitMap[digitSum] > maxGroupSize {
			maxGroupSize = digitMap[digitSum]
		}
	}
	result := 0
	for _, count := range digitMap {
		if count != maxGroupSize {
			continue
		}
		result++
	}
	return result

}
func getCount(num int) int {
	sum := 0
	for num > 0 {
		sum += num % 10
		num = num / 10
	}
	return sum
}

File: LeetCodeProblemsCMake/Problems/LeetCode1399CountLargestGroup.cpp
//
// Created by saeed on 4/23/25.
//

#include "LeetCode1399CountLargestGroup.h"

#include <unordered_map>
using namespace std;
int countDigitSum(int num) {
    int sum{};
    while (num > 0) {
        sum += num % 10;
        num /= 10;
    }
    return sum;
}
int LeetCode1399CountLargestGroup:: countLargestGroup(int n){
    unordered_map<int, int> map{};
    int maxNum{};
    int digitSum{};
    for (int i=1; i<=n; i++) {
        digitSum = countDigitSum(i);
        if (!map.contains(digitSum)) {
            map[digitSum] = 0;
        }
        map[digitSum]++;
        maxNum = max(maxNum, map[digitSum]);
    }
    int result{};
    for (auto & pair : map) {
        if (pair.second != maxNum)
            continue;
        result++;
    }
    return  result;
}
